### PR TITLE
General `Progress` representation for UI Widgets

### PR DESCRIPTION
**Pull request by Weibye**
# Objective

- Both the `SliderWidget` #6236 and the `ProgressBarWidget` #6517 are going to need some representation of progress.
- As we've discussed in those PRs, it would make sense to have both of them use the same underlying representation.

## Solution

- This attempts setting up that underlying and general representation of progress. It is indented to solve the needs of the widgets but ideally would be suitable for additional purposes such as gameplay or loading.
- Additional identified use-case: Getting `Progress` from a `Task` in a `AsyncComputeTaskPool`

---

- ~~Please review and critique both the design but also where to place it. Maybe `bevy_ui` is not the best place for this in the long term?~~
- ~~My intention is that this will be moved over to the `bevy_widgets`-crate once that gets set up (see https://github.com/bevyengine/bevy/pull/6517#pullrequestreview-1171184009) but there might be other, better places for it.~~
- Consensus is that it should live in `bevy_utils`.

### CODE REVIEW

**Comment by Weibye** [comment]
@AxiomaticSemantics, @Pietrek14, review please :)

File: crates/bevy_ui/src/lib.rs [hunk 1]
```diff
@@ -75,6 +78,60 @@ impl Default for UiScale {
     }
 }
 
+/// General representation of progress between two values.
+#[derive(Debug, Clone)]
+pub struct Progress<T>
+where
+    T: Send + Sync + Clone + 'static,
```
Comment (ManevilleF) [inline crates/bevy_ui/src/lib.rs:85]: Why is `T` required to be `'static` ?

File: crates/bevy_ui/src/lib.rs [hunk 2]
```diff
@@ -75,6 +78,60 @@ impl Default for UiScale {
...
+{
+    /// The minimum value that the progress can have, inclusive.
+    pub min: T,
+    /// The maximum value that the progress can have, inlucsive.
+    pub max: T,
+    /// The current value of progress.
+    pub value: T,
+}
+
+impl<T: Send + Sync + Copy> Progress<T>
```
Comment (ManevilleF) [inline crates/bevy_ui/src/lib.rs:95]: You require `T` to implement `Copy` but in the declaration you require `Clone` instead

File: crates/bevy_ui/src/lib.rs [hunk 3]
```diff
@@ -75,6 +78,60 @@ impl Default for UiScale {
...
+where
+    T: PartialOrd<T>,
+{
+    pub fn new(value: T, min: T, max: T) -> Self {
```
Comment (ManevilleF) [inline crates/bevy_ui/src/lib.rs:99]: Should this function check if `value` is actually in range ?

**Review by ManevilleF — COMMENTED** [review]
Interesting tooling, I wonder if having `Num` as a trait bound would be useful, as it probably will always be numeric

File: crates/bevy_ui/src/lib.rs (hunk 3 above)
Comment (Weibye) [reply crates/bevy_ui/src/lib.rs:99]: It does, by calling the `from_range()`, which has the check.

File: crates/bevy_ui/src/lib.rs (hunk 1 above)
Comment (Weibye) [reply crates/bevy_ui/src/lib.rs:85]: Ah, I tried deriving `Reflect`, but not sure if useful or not. Then I forgot to do proper cleanup as a result.

File: crates/bevy_ui/src/lib.rs (hunk 2 above)
Comment (Weibye) [reply crates/bevy_ui/src/lib.rs:95]: Good catch. They should be one or the other, not both I think.

**Comment by Weibye** [comment]
> Interesting tooling, I wonder if having `Num` as a trait bound would be useful, as it probably will always be numeric

That is probably a very reasonable assumption. I can't think of any use-cases where this would not be a number of some kind.

File: crates/bevy_ui/src/lib.rs [hunk 4]
```diff
@@ -75,6 +78,97 @@ impl Default for UiScale {
...
+                min: *range.start(),
+                max: *range.end(),
+            }
+        } else {
+            panic!("Tried creating progress with value out of bounds.");
+        }
+    }
+
+    /// Gets the min bound of the progress.
+    pub fn get_min(&self) -> T {
```
Comment (Pietrek14) [inline crates/bevy_ui/src/lib.rs:124]: Bevy uses `get_` for fallible getters. For getters that can't fail no prefix is used.

File: crates/bevy_ui/src/lib.rs [hunk 5]
```diff
@@ -75,6 +78,97 @@ impl Default for UiScale {
...
+
+    /// Gets the current value of progress.
+    pub fn get_progress(&self) -> T {
+        self.value
+    }
+
+    /// Sets the progress to a new value.
+    ///
+    /// Value must be between `min` and `max`, and will panic otherwise.
+    pub fn set_progress(&mut self, new_value: T) {
```
Comment (Pietrek14) [inline crates/bevy_ui/src/lib.rs:146]: I would rather it return a `Result<()>` than panic.

Comment (Weibye) [reply crates/bevy_ui/src/lib.rs:146]: Is `Result<>` better than a `bool`?

Comment (Pietrek14) [reply crates/bevy_ui/src/lib.rs:146]: I think it signals the intent better.

Comment (Weibye) [reply crates/bevy_ui/src/lib.rs]: Updated to use `Result` instead, and I agree. It is better :)

File: crates/bevy_ui/src/lib.rs [hunk 6]
```diff
@@ -75,6 +78,100 @@ impl Default for UiScale {
...
+        self.value
+    }
+
+    /// Sets the progress to a new value and returns the new value if successful.
+    ///
+    /// Value must be between `min` and `max`.
+    pub fn set_progress(&mut self, new_value: T) -> Result<T, &str> {
+        if self.bounds().contains(&new_value) {
+            self.value = new_value;
+            Ok(self.value)
```
Comment (Pietrek14) [inline crates/bevy_ui/src/lib.rs:149]: I don't think you need to return any value. An `Ok(())` should be enough here.

File: crates/bevy_ui/src/lib.rs [hunk 7]
```diff
@@ -75,6 +78,100 @@ impl Default for UiScale {
...
+        } else {
+            Err("Value outside the bounds of the Progress")
```
Comment (Pietrek14) [inline crates/bevy_ui/src/lib.rs:151]: I would consider making a `ProgressError` enum for errors connected to the `Progress` struct, it's just my preference though, I will still approve it either way.

File: crates/bevy_math/src/lib.rs [hunk 8]
```diff
@@ -22,3 +24,14 @@ pub mod prelude {
 }
 
 pub use glam::*;
+
+/// Maps a value from one range of values to a new range of values.
+pub fn map_range<
```
Comment (alice-i-cecile) [inline crates/bevy_math/src/lib.rs:29]: Should be inlined.

File: crates/bevy_ui/src/lib.rs [hunk 9]
```diff
@@ -75,6 +78,100 @@ impl Default for UiScale {
...
+}
+
+impl<T: Send + Sync + Copy> Progress<T>
+where
+    T: PartialOrd<T>,
+{
+    /// Creates a new progress using a value, min and max.
+    ///
+    /// The value must be within min and max.
+    /// Will panic if value is out-of-bounds.
```
[the remark this answers is not in the document]
Comment (alice-i-cecile) [reply crates/bevy_ui/src/lib.rs:102]: Just a style thing. Same below.

File: crates/bevy_ui/src/lib.rs [hunk 10]
```diff
@@ -75,6 +78,100 @@ impl Default for UiScale {
...
+            Ok(self.value)
+        } else {
+            Err("Value outside the bounds of the Progress")
+        }
+    }
+}
+
+impl Progress<f32> {
+    /// Returns the current progress, normalized between 0 and 1.
+    /// Where 0 represents value == min,
```
Comment (alice-i-cecile) [inline crates/bevy_ui/src/lib.rs:158]:
```suggestion
    ///
    ///  0 represents value == min,
```

File: crates/bevy_ui/src/lib.rs (hunk 7 above)
Comment (alice-i-cecile) [reply crates/bevy_ui/src/lib.rs:151]: Totally agree here.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Some doc nits. Additionally:

1. This should live in `bevy_utils`
2. `Add`, `AddAssign`, `Sub` and `SubAssign` traits.
3. `ProgressError` type.
4. `progress_normalized` -> `normalized` method name.
5. `percent()` constructor method.
6. Derive `Clone` and `PartialEq` on the struct.

File: crates/bevy_ui/src/lib.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (Weibye) [reply crates/bevy_ui/src/lib.rs]: Should this also return a result instead of panicking?

**Comment by Weibye** [comment]
Moved everything to `bevy_utils` and I think I have addressed all comments. feel free to add more :)

File: crates/bevy_utils/src/lib.rs [hunk 11]
```diff
@@ -221,3 +221,197 @@ impl<K: Hash + Eq + PartialEq + Clone, V> PreHashMapExt<K, V> for PreHashMap<K,
...
+    }
+
+    #[test]
+    fn set_value_out_of_bounds() {
+        let min = 0.0;
+        let max = 1.0;
+        let value = 0.5;
+
+        let mut progress = Progress::from_range(value, min..=max).unwrap();
+        assert!(progress.progress() == value);
```
Comment (alice-i-cecile) [inline crates/bevy_utils/src/lib.rs:411]: These should use `assert_eq` for better debug messages.

File: crates/bevy_utils/src/lib.rs [hunk 12]
```diff
@@ -221,3 +221,197 @@ impl<K: Hash + Eq + PartialEq + Clone, V> PreHashMapExt<K, V> for PreHashMap<K,
...
+>(
+    value: T,
+    old_range: (T, T),
+    new_range: (T, T),
+) -> T {
+    (value - old_range.0) / (old_range.1 - old_range.0) * (new_range.1 - new_range.0) + new_range.0
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (alice-i-cecile) [inline crates/bevy_utils/src/lib.rs:353]: I would like to see tests for adding and subtracting over the bound limits.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
There's progress! Some feedback on tests, but once that's done this LGTM.

**Comment by Weibye** [comment]
Additional use-case: Getting `Progress` from a `Task` in a `AsyncComputeTaskPool`

File: crates/bevy_utils/src/lib.rs [hunk 13]
```diff
@@ -221,3 +221,197 @@ impl<K: Hash + Eq + PartialEq + Clone, V> PreHashMapExt<K, V> for PreHashMap<K,
...
+        Self {
+            min: 0.0,
+            max: 1.0,
+            value: 0.0,
+        }
+    }
+}
+
+/// Error types for [`Progress`].
+#[derive(Debug, PartialEq)]
```
Comment (Pietrek14) [inline crates/bevy_utils/src/lib.rs:328]: I think this should use `thiserror` crate for custom error messages.

Comment (alice-i-cecile) [reply crates/bevy_utils/src/lib.rs:328]: Agreed

File: crates/bevy_utils/src/progress.rs [hunk 14]
```diff
@@ -0,0 +1,297 @@
...
+    /// Increases the progress `value` with `rhs`.
+    ///
+    /// # Panic
+    /// Panics if the resulting value is out of bounds.
+    fn add_assign(&mut self, rhs: f32) {
+        if self.set_progress(self.value + rhs).is_err() {
+            panic!("Value of ouf bounds.");
+        }
+    }
+}
```
Comment (Weibye) [inline crates/bevy_utils/src/progress.rs:105]: @alice-i-cecile Do you think this is the correct way to do this? Or should it just clamp the value to the end of the bounds and not panic?

**Comment by Weibye** [comment]
@mockersf Could you please have a look at this from a "general utility across the engine"-perspective?

File: crates/bevy_utils/src/progress.rs (hunk 14 above)
Comment (Weibye) [reply crates/bevy_utils/src/progress.rs:106]: By that you mean the latter of the two options?

Comment (alice-i-cecile) [reply crates/bevy_utils/src/progress.rs:106]: Yes exactly :)

Comment (Weibye) [reply crates/bevy_utils/src/progress.rs]: Changed implementation to be clamped instead of panicking.

**Review by alice-i-cecile — APPROVED** [review]

**Review by Pietrek14 — APPROVED** [review]

File: crates/bevy_utils/src/progress.rs [hunk 15]
```diff
@@ -0,0 +1,302 @@
+use std::ops::{Add, AddAssign, Div, Mul, RangeInclusive, Sub, SubAssign};
+use thiserror::Error;
+
+/// General representation of progress between two values.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Progress<T>
+where
+    T: Send + Sync + Copy + Add + AddAssign + Sub + AddAssign,
```
[the remark this answers is not in the document]
Comment (Weibye) [reply crates/bevy_utils/src/progress.rs:8]: Alright, generic might be a bit overkill for this, I can change it to just f32. I think the adjustable start and end gives a lot of flexibility and utility for various (at least widget) use-cases. 

However, if this only ever should work in `0.0..=1.0` then I agree that the `Progress` struct provides very little value over just a float.

[the remark this answers is not in the document]
Comment (Weibye) [reply crates/bevy_utils/src/progress.rs]: I've removed the generic, but would like to keep the rest of it at the very least for UI purposes :)

**Comment by alice-i-cecile** [comment]
Controversial label added because while this is well-made, it's not entirely clear it's worth the complexity.

**Comment by plof27** [comment]
I took a quick look at this, and honestly it doesn't really seem necessary. I really don't think "what percentage full is this thing" is not a complicated piece of functionality to abstract over. Over in traditional UI, we've been doing just fine without a special construct for progress, since `percentFill: f32` works just fine in nearly all cases.

Additionally, I think this abstraction will quickly start to get applied to too many different use cases to still be meaningful. In my experience, this sort of functionality quickly has its meaning diluted as it's improperly applied to different contexts. Already, I would argue that input sliders and loading bars are two separate areas that don't need to be coupled.

**Comment by mockersf** [comment]
> Already, I would argue that input sliders and loading bars are two separate areas that don't need to be coupled.

Speaking of "progress" for a slider doesn't make much sense...

**Comment by Pietrek14** [comment]
I think this would be useful, else we would have to repeat this logic in `Slider` (#6236) and `ProgressBar` (#6517), which is not the end of the world, but I would much rather have it as a seperate struct. This is a common enough use-case to probably appear in future UI widgets, so rewriting the same thing in all of them could be pretty cumbersome.

As to the name, I feel like calling the slider value "progress" isn't that uncommon, though I don't mind if the name is ultimately changed.

**Comment by Weibye** [comment]
Thanks for all the feedback! Conclusion is to close this and use a normalized float value of "progress" `0.0..=1.0` for the UI elements. 

This can always be reopened if there's a different need where this _is_ the right tool for the job.

**Pull request closed without merging**